File: 05. Searching/Linear Search/linear_search2d.cpp
#include <iostream>

using namespace std;

int main() {
    // Define a 2D array (matrix)
    int rows, cols;
    cout << "Enter the number of rows: ";
    cin >> rows;
    cout << "Enter the number of columns: ";
    cin >> cols;

    int arr[rows][cols];

    // Input elements into the 2D array
    cout << "Enter elements into the 2D array:" << endl;
    for (int i = 0; i < rows; i++) {
        for (int j = 0; j < cols; j++) {
            cin >> arr[i][j];
        }
    }

    // Define the target value you want to search for
    int target;
    cout << "Enter the element you want to search for: ";
    cin >> target;

    // Perform the linear search
    bool found = false;
    for (int i = 0; i < rows; i++) {
        for (int j = 0; j < cols; j++) {
            if (arr[i][j] == target) {
                found = true;
                cout << "Element " << target << " found at position (" << i << ", " << j << ")" << endl;
            }
        }
    }

    if (!found) {
        cout << "Element " << target << " not found in the 2D array." << endl;
    }

    return 0;
}

File: 05. Searching/Linear Search/linear_search2d.py
def linear_search_2d(array, target):
  """Performs a linear search on a 2D array.

  Args:
    array: A 2D array of elements.
    target: The element to search for.

  Returns:
    True if the target element is found in the array, False otherwise.
  """

  for row in range(len(array)):
    for col in range(len(array[row])):
      if array[row][col] == target:
        return True
  return False


# Example usage:

array = [[1, 2, 3], [4, 5, 6], [7, 8, 9]]
target = 5

if linear_search_2d(array, target):
  print("The target element is found in the array.")
else:
  print("The target element is not found in the array.")
